### ISSUE TITLE
We should implement undo.

### ISSUE THREAD

**Issue by skx** [skx/kilua#14]
This should be pretty simple..

**Comment by skx** [comment]
To give more detail this should actually be pretty simple:
- We create a stack of "undo" operations.
- When the function `undo_lua` is called we just execute the action.
  - Obviously we bind `Ctrl-z` to `undo()`.

So what are actions?  
- If the user inserts a character then the undo for that is "delete a character".
- If the user deletes a character then the undo for that is "insert character XX"
- If the user moves the cursor down the undo for that is "move cursor up".

So we need a stack that allows:
- move
- insert
- delete

Then we just need to record the direction (for moves) and the character (for insertions).

**Comment by skx** [comment]
NOTE: We should throw away the undo stack when a) saving b) loading a new file.

**Comment by skx** [comment]
Perhaps this isn't as easy as I'd first thought.

The specific problem I have is that if the user presses `find` (`ctrl-f`) their cursor warps, and we cannot start undoing things - because the  undo-history attempts to keep track of "insert", "delete", and "move".  It'll replay things, but in the wrong location.

So the next obvious solution is to drop remembering the idea of `move` and only track two things:
- character inserted at x,y.
- position x,y deleted.

That's trivial and should be fine.  However the case of undoing those things becomes harder instead:
- If a character is inserted the undo-operation becomes:
  - Move to x,y
  - delete.
- If a character is deleted the undo-operation becomes:
  - Move to x,y
  - Insert.

However the insertion will most likely fail.  Consider the case where you have a line of text:

```
  this is my lin[e]
```

`[e]` is deleted, and needs to be inserted.  You cannot move to the position it used to occupy, because it doesn't exist in `row->chars`/`row->render`.  So you have to move to the position left-of it.  That might work, but it feels hacky.

Finally a third option is to don't track "insert"/"delete".  Instead track `erow` objects:
- If an operation is carried out on an `erow`
  - Copy that row to the stack, along with its offset.

That makes addition/removal simple to manage.  But it complicates the case of row-deletion.  If a row is deleted we need to know where to replace it to on undo, and obviously all row indexes are going to be relative.

In conclusion this needs more thought.

**Comment by skx** [comment]
Think we're good.  Report bugs if you see them :)
